### PR TITLE
Vaults

### PR DESCRIPTION
**Pull request by matzayonc**

### CODE REVIEW

File: src/vaults.ts [hunk 1]
```diff
@@ -1,59 +1,225 @@
...
+
+    const neededAmount = toDecimal(
+      amount.val.muln(102).divn(100).sub(syntheticAccount.amount),
+      amount.scale
+    )
+
+    // Minimum amount that can be traded on synthetify
+    const swapAmount = neededAmount.val.gt(new BN(1000)) ? neededAmount.val : new BN(1001)
+
+    if (swapAmount.gt(xUSDAccount.amount)) throw new Error('not enough xUSD')
```
Comment (NorbertBodziony) [inline src/vaults.ts:117]: We should liquidate with all we have left.

File: src/vaults.ts [hunk 2]
```diff
@@ -1,59 +1,225 @@
...
+    // Swap to the right synthetic
+    if (!isUsdTheSynthetic && neededAmount.val.gten(0)) {
+      console.log('Swapping synthetics..')
+
+      tx.add(
+        Token.createApproveInstruction(
+          TOKEN_PROGRAM_ID,
+          xUSDAccount.address,
+          exchange.exchangeAuthority,
+          wallet.publicKey,
```
Comment (NorbertBodziony) [inline src/vaults.ts:127]: What if our account on token does not exist ?

**Review by NorbertBodziony — CHANGES_REQUESTED** [review]

File: src/vaults.ts (hunk 2 above)
Comment (matzayonc) [reply src/vaults.ts:127]: on line 143 all of them are created

File: src/vaults.ts (hunk 1 above)
Comment (matzayonc) [reply src/vaults.ts:117]: will work like that in the final version

**Review by wojciech-cichocki — APPROVED** [review]

**Review by NorbertBodziony — APPROVED** [review]

**Pull request merged**